### DISCUSSION TITLE
Show the number of cycles used to execute a call

### DISCUSSION THREAD

**Discussion by ghost** [dfinity/sdk#2783] [Feature Requests]
This came from a discussion that took place with the ICP.Labs team last week when they visited the Zurich office. The idea is that dfx should return the number of cycles a call consumed.

There are still a few open questions, and I'm leaving this intentionally vague to encourage some brainstorming and creativity.

1. does this assume we want to return cycles consumed by canister methods only?
2. how does dfx access this data? we have the ability to profile cycles within canisters, but I'm not aware of any way to currently expose that to external callers. It seems like dfx would need to instrument the WASM before deploying, which I'm not a huge fan of.

**Comment by domwoe** [comment]
cc @lastmjs

**Comment by viviveevee** [comment]
I think this would require instrumenting the WASM or replica support

**Reply by chenyan-dfinity** [reply]
We can call `canister_status` before and after the call to get the cycle consumption. Some limitations, which also applies to the instrumentation method:

* Doesn't work for query method
* Doesn't work for concurrent calls
* Doesn't work in the presence of heartbeat
* When the method is making an inter-canister call, it only counts the cycle consumed from the caller. We need a way to get the callee cost. (This is the main reason I don't like to report cycles after each call: we are missing the callee cost).

**Reply by viviveevee** [reply]
Doesn't calling `canister_status` also cost cycles since it's an update call, too? And if so, is that cost (somewhat) deterministic? That could mess with the results.

**Comment by lastmjs** [comment]
At a high level, I think the idea is very similar to how Ethereum works. When you perform a transaction, you know if the exact gas costs afterwards.

I want to execute a call and know exactly how much it cost my canister...maybe bonus points if it gives more info about other canisters.
